R/db-helpers.R

```diff
@@ -167,43 +167,55 @@ repair_table_names_for_db <- function(table_names, temporary, con) {
   quote_ids(names, con)
 }
 
-get_src_tbl_names <- function(src, schema = NULL) {
-  con <- src$con
-
-  if (is_null(schema)) {
-    if (!is_mssql(src) && !is_postgres(src)) {
-      # `src_tbls()` returns system tables and tables in other schemas than default schema only for MSSQL
-      return(src_tbls(src))
-    } else if (is_mssql(src)) {
-      # MSSQL
-      schema <- "dbo"
-    } else {
-      # Postgres
-      schema <- "public"
-    }
-  } else if (!is_mssql(con) && !is_postgres(con)) {
-    warn("Argument 'schema' ignored: currently only supports MSSQL and Postgres")
+get_src_tbl_names <- function(src, schema = NULL, dbname = NULL) {
+  if (!is_mssql(src) && !is_postgres(src)) {
+    warn_if_not_null(schema)
+    warn_if_not_null(dbname, only_on = "MSSQL")
     return(src_tbls(src))
   }
 
-  # src is now either Postgres or MSSQL, schema is not `NULL`
+  con <- src$con
+
   if (is_mssql(src)) {
```

Comment:
Can you please extract functions for the code in the `if {}`, and also be explicit about the database in the check (if MSSQL, if Postgres)?